src/List.h

```diff
@@ -20,67 +20,353 @@
+		}
 
 	int MemoryAllocation() const
-		{ return padded_sizeof(*this) + pad_size(max_entries * sizeof(ent)); }
+		{ return padded_sizeof(*this) + pad_size(max_entries * sizeof(T)); }
 
-protected:
-	~BaseList()		{ free(entry); }
-	explicit BaseList(int = 0);
-	BaseList(const BaseList&);
-	BaseList(BaseList&&);
-	BaseList(const ent* arr, int n);
+	void insert(T a)	// add at head of list
+		{
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
 
-	BaseList& operator=(const BaseList&);
-	BaseList& operator=(BaseList&&);
+		for ( int i = num_entries; i > 0; --i )
+			entries[i] = entries[i-1];	// move all pointers up one
 
-	void insert(ent);	// add at head of list
+		++num_entries;
+		entries[0] = a;
+		}
 
 	// Assumes that the list is sorted and inserts at correct position.
-	void sortedinsert(ent, list_cmp_func cmp_func);
+	void sortedinsert(T a, list_cmp_func cmp_func)
+		{
+		// We optimize for the case that the new element is
+		// larger than most of the current entries.
+
+		// First append element.
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
+
+		entries[num_entries++] = a;
+
+		// Then move it to the correct place.
+		T tmp;
+		for ( int i = num_entries - 1; i > 0; --i )
+			{
+			if ( cmp_func(entries[i],entries[i-1]) <= 0 )
+				break;
 
-	void append(ent);	// add to end of list
-	ent remove(ent);	// delete entry from list
-	ent remove_nth(int);	// delete nth entry from list
-	ent get();		// return and remove ent at end of list
-	ent last()		// return at end of list
-		{ return entry[num_entries-1]; }
+			tmp = entries[i];
+			entries[i] = entries[i-1];
+			entries[i-1] = tmp;
+			}
+		}
+
+	void append(T a)	// add to end of list
+		{
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
+
+		entries[num_entries++] = a;
+		}
+
+	T remove(T a)	// delete entry from list
+		{
+		int i;
+		for ( i = 0; i < num_entries && a != entries[i]; ++i )
+			;
+
+		return remove_nth(i);
+		}
+
+	T remove_nth(int n)	// delete nth entry from list
+		{
+		if ( n < 0 || n >= num_entries )
+			return 0;
+
+		T old_ent = entries[n];
+		--num_entries;
+
+		for ( ; n < num_entries; ++n )
+			entries[n] = entries[n+1];
+
+		entries[n] = nullptr;	// for debugging
+		return old_ent;
+		}
+
+	T get()		// return and remove ent at end of list
+		{
+		if ( num_entries == 0 )
+			return 0;
+
+		return entries[--num_entries];
+		}
+
+	T last()		// return at end of list
+		{ return entries[num_entries-1]; }
 
 	// Return 0 if ent is not in the list, ent otherwise.
-	ent is_member(ent) const;
+	bool is_member(T a) const
+		{
+		int pos = member_pos(a);
+		return pos != -1;
+		}
 
 	// Returns -1 if ent is not in the list, otherwise its position.
-	int member_pos(ent) const;
+	int member_pos(T e) const
+		{
+		int i;
+		for ( i = 0; i < length() && e != entries[i]; ++i )
```

Comment:
Might be a little cleaner to use `std::find(begin(this), end(this), e)`

Reply:
If you need to use the index instead of the iterator:
```c++
return std::distance(begin(this), std::find(begin(this), end(this), e));
```

Reply:
It'd be nicer to just have Ranges support so I could do this in-place.